### PR TITLE
Shared library

### LINKED ISSUE #621: Shared libraries?
Hi! I maintain the Fedora package for Box2D. I'm looking at upgrading from 2.3.1 to 2.4.0, but I can't get the new cmake build system to build shared libraries, only static. Is this by design, or have I missed something? Thank you!

### PR DESCRIPTION
**Pull request by jube**
This pull request add support for building Box2D as a shared library. It fixes #621. By default, the library is built as a static library, you have to add `-DBUILD_SHARED_LIBS=ON` to the cmake invocation to build the shared library.

If you merge #630 at the same time, you will have to uncomment some lines in `src/CMakeLists.txt` so that the library can work properly once installed. Anyway, merging both should result in conflicts. If you plan to merge both, merge #630 first, and I will make the necessary modifications to this PR so that it can be merged smoothly.

### CODE REVIEW

**Comment by erincatto** [comment]
I think we would need some test to make sure this works. It would be cool if the testbed optionally can run Box2D as a DLL/shared lib.

**Comment by jube** [comment]
I have tested both configurations (shared and static) and the testbed works in both cases. I work on a Debian 10 (stable).

That's why I had to declare some variables that were not declared in headers but that were used in the testbed. With the `CMAKE_CXX_VISIBILITY_PRESET hidden`, all symbols that are not properly exported are not visible (that's the default behavior on Windows, but not on Linux). So I am pretty confident it works well with Box2D as a shared library.

If someone else can try this patch, it would be nice. Maybe @limburgher?

If you want to test it on Windows yourself, you have to run `cmake-gui` in the build folder of Box2D and check the `BUILD_SHARED_LIBS` option (and then, press `Configure` and `Generate`). Normally, it will trigger a whole reconfiguration/recompile of the library in Visual Studio. I will be able to test this patch on Windows tonight maybe.

**Comment by svenstaro** [comment]
FWIW I tested and approve these changes. I'd appreciate if they were upstreamed.

File: CMakeLists.txt
Comment (bschatt) [inline CMakeLists.txt:12]: You don't need that. This is a global cmake flag.

File: CMakeLists.txt
Comment (bschatt) [inline CMakeLists.txt:10]: You should probably add `set(CMAKE_VISIBILITY_INLINES_HIDDEN 1)` as well.

File: include/box2d/b2_collision.h
Comment (bschatt) [inline include/box2d/b2_collision.h:26]: is this needed? Doesn't seem related to the current PR, right?

File: include/box2d/b2_growable_stack.h
Comment (bschatt) [inline include/box2d/b2_growable_stack.h:26]: Doesn't seem related to the current PR.

File: include/box2d/b2_growable_stack.h
Comment (bschatt) [inline include/box2d/b2_growable_stack.h:60]: I think this should be reverted too.

File: include/box2d/b2_math.h
Comment (bschatt) [inline include/box2d/b2_math.h:26]: Same thing here.

File: include/box2d/b2_math.h
Comment (bschatt) [inline include/box2d/b2_math.h:34]: Not going to mention all of them in this file.

File: include/box2d/b2_settings.h
Comment (bschatt) [inline include/box2d/b2_settings.h:28]: Not related to the current PR.

File: src/CMakeLists.txt
Comment (bschatt) [inline src/CMakeLists.txt:107]: You can use [CMAKE_INCLUDE_CURRENT_DIR](https://cmake.org/cmake/help/v3.0/variable/CMAKE_INCLUDE_CURRENT_DIR.html) instead.

File: src/CMakeLists.txt
Comment (bschatt) [inline src/CMakeLists.txt:134]: CMake defines this by default when you compile a shared lib.

File: src/CMakeLists.txt
Comment (bschatt) [inline src/CMakeLists.txt:122]: I tried it and it generated this file:
```
#ifndef B2_API_H
#define B2_API_H

#ifdef B2_STATIC
#  define B2_API
#  define B2_NO_EXPORT
#else
#  ifndef B2_API
#    ifdef box2d_EXPORTS
        /* We are building this library */
#      define B2_API __attribute__((visibility("default")))
#    else
        /* We are using this library */
#      define B2_API __attribute__((visibility("default")))
#    endif
#  endif

#  ifndef B2_NO_EXPORT
#    define B2_NO_EXPORT __attribute__((visibility("hidden")))
#  endif
#endif

#ifndef B2_DEPRECATED
#  define B2_DEPRECATED __attribute__ ((__deprecated__))
#endif

#ifndef B2_DEPRECATED_EXPORT
#  define B2_DEPRECATED_EXPORT B2_API B2_DEPRECATED
#endif

#ifndef B2_DEPRECATED_NO_EXPORT
#  define B2_DEPRECATED_NO_EXPORT B2_NO_EXPORT B2_DEPRECATED
#endif

#if 0 /* DEFINE_NO_DEPRECATED */
#  ifndef B2_NO_DEPRECATED
#    define B2_NO_DEPRECATED
#  endif
#endif

#endif /* B2_API_H */
```

I think this is a breaking change on Windows. Every project using static box2d lib will have to define `B2_STATIC`. CMake projects won't be affected since you added it to the public include paths. Everyone else will have to define it manually.

Also people importing box2d inside their source tree and not compiling with CMake will have to generate this file manually.

The solution to both problem would be to create the file yourself and add it to the repo. Instead of using `#if B2_STATIC` you could use `#ifndef B2_DYNAMIC`.

File: include/box2d/b2_contact.h
Comment (bschatt) [inline include/box2d/b2_contact.h:355]: I think it would be cleaner to add `B2_API` to the declaration.

File: include/box2d/b2_distance.h
Comment (bschatt) [inline include/box2d/b2_distance.h:172]: I think it would be cleaner to add B2_API to the declaration.

File: include/box2d/b2_time_of_impact.h
Comment (bschatt) [inline include/box2d/b2_time_of_impact.h:68]: I think it would be cleaner to add B2_API to the declarations.

**Review by bschatt — COMMENTED** [review]
I don't think it's a good idea to expose internal variables as extern. I'm also a bit worried about the b2_api header generation. Apart from that, it's all good. Nice job :+1:

File: CMakeLists.txt
Comment (jube) [reply CMakeLists.txt:12]: The [documentation](https://cmake.org/cmake/help/v3.8/variable/BUILD_SHARED_LIBS.html) says that it's often an option.

File: CMakeLists.txt
Comment (jube) [reply CMakeLists.txt:10]: Many functions are defined inline in the headers, I do not think it's a good idea to make them hidden by default. In fact, all inline functions are public, there is no private inline.

File: include/box2d/b2_collision.h
Comment (jube) [reply include/box2d/b2_collision.h:26]: It's a bonus, headers in `.h` have been deprecated for years. They may not be available in a standard compiler (even if the may compilers make them available). It's a good practice to use C++ version of C headers.

File: src/CMakeLists.txt
Comment (jube) [reply src/CMakeLists.txt:107]: Right!

File: src/CMakeLists.txt
Comment (jube) [reply src/CMakeLists.txt:134]: Right!

File: src/CMakeLists.txt
Comment (jube) [reply src/CMakeLists.txt:122]: If you use cmake, it's not a breaking change, the default behavior is still the same as before. If you want to use box2d, you have to compile it cleanly with cmake. If you want to do garbage when compiling cmake, you will have garbage at the end. That's a cheap price to pay to have many configurations available (shared/static in this case) for everybody.

The `b2_api.h` is generated and should be generated. It's a CMake feature and it's very useful! People wanting to compile box2d in a weird way (not with cmake) will have to create it themselves, it's not a box2d problem.

File: include/box2d/b2_contact.h
Comment (jube) [reply include/box2d/b2_contact.h:355]: Yes, good idea! It will be better.

File: src/CMakeLists.txt
Comment (bschatt) [reply src/CMakeLists.txt:122]: If you *use* box2d in a project that doesn't use cmake (or even if you're using an old FindBox2d cmake module, [a lot of projects are shipping their own](https://www.google.com/search?q=findbox2d.cmake)) you now have to define B2_STATIC manually while it wasn't required before. That's a breaking change.

As for the header I think it's a shame to use code generation for such a simple macro. Most projects don't do that. Gcc [documentation](https://gcc.gnu.org/wiki/Visibility) provides an example.

File: include/box2d/b2_collision.h
Comment (bschatt) [reply include/box2d/b2_collision.h:26]: A few years ago I had some issues with box2d because it was using some of these "modern" headers (on a proprietary platform). Changing that might break compilation on some platforms and doesn't add much anyway.

File: CMakeLists.txt
Comment (bschatt) [reply CMakeLists.txt:10]: Yes, that's exactly why you'd want to enable this option. Otherwise the inline functions might get exported. Those aren't needed since they are inline (i.e. the code will be available to consuming projects).

File: CMakeLists.txt
Comment (bschatt) [reply CMakeLists.txt:12]: Right. But that doesn't mean it's needed. It will work in the same way, with or without this line, that's all I'm saying :)

File: include/box2d/b2_collision.h
Comment (jube) [reply include/box2d/b2_collision.h:26]: Non-conformant compilers should die.

File: src/CMakeLists.txt
Comment (jube) [reply src/CMakeLists.txt:122]: Well, the new version of Box2D changed the name of the headers, which is a bigger breaking change and already breaks all the existing `FindBox2D`. Time to move on and use the other patch.

File: CMakeLists.txt
Comment (jube) [reply CMakeLists.txt:12]: Having an option allows to set a default value (`OFF`) which is the default now.

File: src/CMakeLists.txt
Comment (bschatt) [reply src/CMakeLists.txt:122]: You're right. I think think it's a shame to rely on code generation for such a simple thing but it works :+1:

**Review by bschatt — APPROVED** [review]
Nice :+1:

**Comment by jube** [comment]
I just resolved the merge conflict after #630 has been merged in master.

**Comment by svenstaro** [comment]
This needs a rebase again.

**Comment by jube** [comment]
Rebased. Hope it works.

**Comment by svenstaro** [comment]
@erincatto can this be merged? It would address my downstream maintainer needs. :)

**Comment by erincatto** [comment]
The diffs are showing changes from other PRs. Is there a way to clean this up or view the diffs of just this PR?

**Comment by jube** [comment]
My bad, I messed up with the latest rebase. I will try to fix it as soon as possible.

@erincatto you still have to decide about `b2_api.h`. In this PR, the file is generated by cmake at configuration time. It is then installed in the include directory next to the other files. The other option would be to define this file directly in the repository. You can see the discussion with @bschatt above. I find it easier to let cmake generate it but I can understand @bschatt's arguments and I could change the PR if necessary. It's up to you.

**Comment by jube** [comment]
This time, it should be ok.

**Comment by erincatto** [comment]
Thanks. I will test this soon.

**Comment by erincatto** [comment]
Tested this finally. It looks like box2d.dll is not in the path of testbed.exe
image

**Comment by erincatto** [comment]
It looks like you are using CMake to generate a file that must be included, b2_api.h. This goes against my goal of people being able to build Box2D without CMake.

**Comment by jube** [comment]
`b2_api.h` is not generated anymore by CMake, a file has been created with the same features in the `include/box2d` directory.

As for the other problem, the solution is to put all the binaries (executables and libraries) in the same build folder. So now, everything is built in a `bin` folder in the build directory. The only exception is the `unit_test` binary as it has been put in the build directory manually. But I think it should be put in the same `bin` directory (delete the lines 17-18 in `unit-test/CMakeLists.txt`).

**Comment by erincatto** [comment]
@jube thanks. Please modify the unit test too as you mentioned.

**Comment by jube** [comment]
Done! I handled the `B2_SHARED` too.

I think it would be better to squash all the commits before merging to avoid pollution in the main branch.

File: CMakeLists.txt
Comment (Nek-) [inline CMakeLists.txt:18]: This option is duplicated!

**Comment by Nek-** [comment]
Also isn't the install script missing @jube ?

File: CMakeLists.txt
Comment (jube) [reply CMakeLists.txt:18]: You are right!

**Comment by jube** [comment]
@Nek- Install instruction are spread in `CMakeLists.txt` and `src/CMakeLists.txt`. It seems to work. What is missing?

**Comment by Nek-** [comment]
@jube idk I'm not an expert in C++ and Cmake 😅 I just was not able to run the installation.

**Review by erincatto — APPROVED** [review]
Looks good! Thanks for addressing all the feedback.

**Comment by svenstaro** [comment]
Any chance for a new release including this change? :)

**Comment by erincatto** [comment]
@svenstaro I'm working on that right now

**Pull request merged**
